connectors/sources/graphql.py

```diff
@@ -0,0 +1,333 @@
+                "order": 6,
+                "sensitive": True,
+                "type": "str",
+            },
+            "graphql_query": {
+                "display": "textarea",
+                "label": "GraphQL Body",
+                "order": 7,
+                "type": "str",
+            },
+            "graphql_variables": {
+                "depends_on": [{"field": "http_method", "value": "post"}],
+                "display": "textarea",
+                "label": "Graphql Variables",
+                "order": 8,
+                "type": "str",
+                "required": False,
+            },
+            "object_list": {
+                "label": "GraphQL Objects List",
+                "order": 9,
+                "type": "list",
+            },
+            "headers": {
+                "label": "Headers",
+                "order": 10,
+                "type": "str",
+                "required": False,
+            },
+        }
+
+    def is_query(self, graphql_query):
+        ast = parse(graphql_query)  # pyright: ignore
+        operation_type = ast.definitions[0].operation  # pyright: ignore
+        if operation_type.value == "query":
+            return True
+        return False
+
+    async def validate_config(self):
+        """Validates whether user input is empty or not for configuration fields
+        Also validate, if user configured repositories are accessible or not and scope of the token
+        """
+        await super().validate_config()
+        if self.is_query(graphql_query=self.graphql_client.graphql_query) is False:
+            msg = "Mutation is not supported."
+            raise ConfigurableFieldValueError(msg)
+
+        headers = self.graphql_client.configuration["headers"]
+        graphql_variables = self.graphql_client.configuration["graphql_variables"]
+        if headers:
+            try:
+                self.graphql_client.headers = json.loads(headers)
+            except Exception as exception:
+                msg = f"Error while processing headers. Exception: {exception}"
+                raise ConfigurableFieldValueError(msg) from exception
+
+        if graphql_variables:
+            try:
+                self.graphql_client.variables = json.loads(graphql_variables)
+            except Exception as exception:
+                msg = f"Error while processing variables. Exception: {exception}"
+                raise ConfigurableFieldValueError(msg) from exception
+
+    async def close(self):
+        await self.graphql_client.close()
+
+    async def ping(self):
+        try:
+            await self.graphql_client.ping()
+            self._logger.debug("Successfully connected to GraphQL Instance.")
+        except Exception:
+            self._logger.exception("Error while connecting to GraphQL Instance.")
+            raise
+
+    async def fetch_data(self, graphql_query):
+        data = await self.graphql_client.make_request(graphql_query=graphql_query)
```

Comment:
What would happen if our graphql backend has e.g. 100000000 docs? Would `make_request` result in a huge response containing all docs? We should paginate/scroll through the graphql db to actually make use of async programming benefits

Reply:
In the connector user is able to execute one query only and almost every source have some GraphQL limit to return objects (for example first: 100 object). So connector only fetches objects which return by single query. So i guess pagination is not required.

Reply:
Can we modify the user configuration? Suppose user configure below query:
```
repositories (first: 100, after:"endCursor"){
                  name
                  stars
}
```
So, in above query user explicitly mentioned that he/she wants 100 objects only.

Also, I guess pagination might be different in different source. So, how we can modify the query?

Reply:
> So, in above query user explicitly mentioned that he/she wants 100 objects only.

What if they want all the objects? We need to add a support to be able to synchronize _all_ data matching a user query between the graphql instance and the index. 

> Also, I guess pagination might be different in different source. So, how we can modify the query?
What problematic case do you have in mind? We need to have a strategy for adding pagination to graphql queries, perhaps there are some open-source examples how to do it?

Reply:
> What problematic case do you have in mind?

For example, we are working with 2 sources, `https://api.github.com/graphql` and `https://countries.trevorblades.com/graphql`. So, GitHub GraphQL supports pagination using `pageInfo` and `endCursor`. 
query: 
```
query ($login: String!, $cursor: String) {{
    user(login: $login) {
        repositories(first: 100, after: $cursor) {
        pageInfo {
            hasNextPage
            endCursor
        }
        nodes {
            id
            updatedAt
            name
            nameWithOwner
        }
        }
    }
    }
```
So above query fetches 100 repositories in each page and using `endCursor` we can move forward and fetches all the repositories.

But for `countries.trevorblades.com` we do not have `pageInfo` available and also does not have keywords like `first` , `after` etc. Using below query we can fetch all the countries using single API call. So, in generic how we can add pagination for all the sources which supports GraphQL?

```
query Countries {
    countries {
        name
    }
}
```

Reply:
Hey @parthpuri-elastic 

> So, in generic how we can add pagination for all the sources which supports GraphQL?

I played with `countries.trevorblades.com` and yeah this doesn't support pagination at all, so in that case there is nothing we can do. 

Loooking at docs it seems that usually servers support:
- Cursor-based pagination with variables:
  - `$afterCursor`: The cursor to fetch items after. Initially, this could be null to start from the beginning.
  - `$pageSize`: The number of items to fetch.
- Offset-based pagination with variables
  - `$offset`: The offset from the beginning of the dataset to fetch items from. Initially, this could be 0.
  - `$limit`: The maximum number of items to return. 

I think we should let the user define the query with pagination variables. The connector implementation should use those variables to paginate, so:
- after each page load set `afterCursor` to `endCursor`
- after each page load set `offset` to (`offset + limit`)

What do you think about this approach? Our aim is to keep the connector implementations simple while supporting paginating over data if users wants it, what do you think? 😄

Reply:
Sure, we can move forward with this approach for now 👍